### PR TITLE
Add meta and signature validation to RPM dist in Validation workflow

### PR DESCRIPTION
**Pull request by Divyaasm**
### Description
Migrate RPM metadata validation and signature validation from groovy to python

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/4968
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/validation_workflow/rpm/validation_rpm.py [hunk 1]
```diff
@@ -68,3 +70,62 @@ def cleanup(self) -> bool:
...
+        for key, value in ref_map.items():
+            if key == "Architecture":
+                if value == 'x64':
+                    assert meta_map.get(key) == 'x86_64'
+                elif value == 'arm64':
+                    assert meta_map.get(key) == 'aarch64'
+            else:
+                if meta_map.get(key) == value:
+                    logging.info(value)
+            logging.info(f"Meta data for {key} is validated")
```
Comment (peterzhuamazon) [inline src/validation_workflow/rpm/validation_rpm.py:112]: A bit confused here as you assert architecture, but for other fields you just did a logging if exist?
What if not exist should you assert those as well?

File: src/validation_workflow/rpm/validation_rpm.py [hunk 2]
```diff
@@ -68,3 +70,62 @@ def cleanup(self) -> bool:
...
+        (_, stdout, _) = execute(f'rpm -K -v {os.path.join(self.tmp_dir.path, self.filename)}', ".")
+        logging.info(stdout)
+        key_list = ["Header V4 RSA/SHA512 Signature, key ID 9310d3fc", "Header SHA256 digest", "Header SHA1 digest", "Payload SHA256 digest", "V4 RSA/SHA512 Signature, key ID 9310d3fc", "MD5 digest"]
+        present_key = []
+        for line in stdout.rstrip('\n').split('\n'):
+            key = line.split(':')[0].strip()
+            if key != os.path.join(self.tmp_dir.path, self.filename):
+                if "OK" == line.split(':')[1].strip():
+                    logging.info(f"{key} is validated as: {line}")
+                    present_key.append(key)
```
Comment (peterzhuamazon) [inline src/validation_workflow/rpm/validation_rpm.py:126]: Same here, what if OK is not exist here.
Will you assert?

File: src/validation_workflow/rpm/validation_rpm.py [hunk 3]
```diff
@@ -68,3 +70,62 @@ def cleanup(self) -> bool:
...
+        logging.info("Validation of all key digests starts: ")
+        for digest in key_list:
+            if digest in present_key:
+                logging.info(f'Key digest "{digest}" is validated to be present.')
+        logging.info("Validation for signature of RPM distribution completed.")
```
Comment (peterzhuamazon) [inline src/validation_workflow/rpm/validation_rpm.py:131]: Same as above.

**Review by peterzhuamazon — CHANGES_REQUESTED** [review]
Hi @Divyaasm please check the comments here.

Thanks.

File: src/validation_workflow/rpm/validation_rpm.py (hunk 2 above)
Comment (Divyaasm) [reply src/validation_workflow/rpm/validation_rpm.py:126]:
````
/tmp/tmphyiftiqz/opensearch-2.17.0-linux-x64.rpm:

    Header V4 RSA/SHA512 Signature, key ID 9310d3fc: OK

    Header SHA256 digest: OK

    Header SHA1 digest: OK

    Payload SHA256 digest: OK

    V4 RSA/SHA512 Signature, key ID 9310d3fc: OK

    MD5 digest: OK
````

This is the expected output. 
We are checking if all the listed keys
        key_list = ["Header V4 RSA/SHA512 Signature, key ID 9310d3fc", "Header SHA256 digest", "Header SHA1 digest", "Payload SHA256 digest", "V4 RSA/SHA512 Signature, key ID 9310d3fc", "MD5 digest"]
 are present in the output.

File: src/validation_workflow/rpm/validation_rpm.py (hunk 1 above)
Comment (Divyaasm) [reply src/validation_workflow/rpm/validation_rpm.py:112]: We are comparing each item with the result with predefined lists. 
if meta_map.get(key) == value: 
will do the same this assert meta_map.get(key) == 'x86_64'
I just missed the logging statement saying architecture is validated.

File: src/validation_workflow/rpm/validation_rpm.py (hunk 3 above)
Comment (Divyaasm) [reply src/validation_workflow/rpm/validation_rpm.py:131]: Have done the exact thing whats groovy libraries does currently.
https://github.com/opensearch-project/opensearch-build-libraries/blob/f0454a47610c96e5840637a758205a86f3428a39/vars/rpmMetaValidation.groovy#L26

File: src/validation_workflow/rpm/validation_rpm.py (hunk 2 above)
Comment (peterzhuamazon) [reply src/validation_workflow/rpm/validation_rpm.py:126]: Yeah, but what if a rpm is invalid and the digest is not `OK` in return?

File: src/validation_workflow/rpm/validation_rpm.py (hunk 3 above)
Comment (peterzhuamazon) [reply src/validation_workflow/rpm/validation_rpm.py:131]: I was saying you did not check what if digest not present.

File: tests/tests_validation_workflow/test_validation_rpm.py [hunk 4]
```diff
@@ -207,3 +209,51 @@ def test_cleanup(self, mock_temporary_directory: Mock, mock_validation_args: Moc
 
         result = validate_rpm.cleanup()
         self.assertTrue(result)
+
```
Comment (peterzhuamazon) [inline tests/tests_validation_workflow/test_validation_rpm.py:212]: Can you also add two test cases for `invalid` metadata and signature?

Thanks.

File: src/validation_workflow/rpm/validation_rpm.py [hunk 5]
```diff
@@ -37,12 +39,7 @@ def start_cluster(self) -> bool:
         try:
             for project in self.args.projects:
                 execute(f'sudo systemctl start {project}', ".")
-                (stdout, stderr, status) = execute(f'sudo systemctl status {project}', ".")
-                if(status == 0):
-                    logging.info(stdout)
-                else:
-                    logging.info(stderr)
-
+                (status, _, _) = execute(f'sudo systemctl status {project}', ".")
```
Comment (peterzhuamazon) [inline src/validation_workflow/rpm/validation_rpm.py:42]: Why change this?

File: src/validation_workflow/rpm/validation_rpm.py [hunk 6]
```diff
@@ -68,3 +65,65 @@ def cleanup(self) -> bool:
...
+
+    def validate_signature(self) -> None:
+        (_, stdout, _) = execute(f'rpm -K -v {os.path.join(self.tmp_dir.path, self.filename)}', ".")
+        logging.info(stdout)
+        key_list = ["Header V4 RSA/SHA512 Signature, key ID 9310d3fc", "Header SHA256 digest", "Header SHA1 digest",
+                    "Payload SHA256 digest", "V4 RSA/SHA512 Signature, key ID 9310d3fc", "MD5 digest"]
+        present_key = []
+        for line in stdout.rstrip('\n').split('\n'):
+            key = line.split(':')[0].strip()
+            if key != os.path.join(self.tmp_dir.path, self.filename):
```
Comment (peterzhuamazon) [inline src/validation_workflow/rpm/validation_rpm.py:119]: What's this for? Could you give the stdout example to understand?
Thanks.

Comment (Divyaasm) [reply src/validation_workflow/rpm/validation_rpm.py:124]: The output is similar to:
```
/tmp/tmpx80rtubp/opensearch-2.16.0-linux-x64.rpm:

    Header V4 RSA/SHA512 Signature, key ID 9310d3fc: OK

    Header SHA256 digest: OK

    Header SHA1 digest: OK

    Payload SHA256 digest: OK

    V4 RSA/SHA512 Signature, key ID 9310d3fc: OK

    MD5 digest: OK
```
It ignores the first line here.

Comment (peterzhuamazon) [reply src/validation_workflow/rpm/validation_rpm.py:124]: Are you able to just remove the first line always when you receive the output?
It is not very pretty to check if it starts on `/tmp/<>/opensearch`.
Or at least just check if it contains filename should be enough.

**Review by peterzhuamazon — APPROVED** [review]

**Pull request merged**